### DISCUSSION TITLE
KeyBinding does not support Click events

### DISCUSSION THREAD

**Discussion by cornerbowlsoftware** [AvaloniaUI/Avalonia#16313] [General]
IMHO proper MVVM means the VM has no concept of the platform it is running.  In my case my VMs are in a .Net Core Library then reused on all platforms. Since some event handlers need to set focus to other controls, for example when pressing the Delete button to delete an item from a list, code behind must be used.  A work around it to monitor the DataContext then assign a callback function to the VM's command handler, however, when sharing components between views this is problematic.  KeyBindings should support both Command and Click options allowing both scenarios, MVVM and code behind.

**Comment by stevemonaco** [comment]
AFAIK, `KeyBinding` exists mostly as an MVVM adapter so events (particularly event arguments) don't leak into the VM. You can still bind `Command` to `ICommand`s or methods in a View's code-behind, However, since `KeyBinding`s aren't attached to a control tree IIRC, you can't use `RelativeSource` / `$parent` bindings, but `ElementName` type bindings are still valid. eg.

```xml
<Window x:Name="window" ....>
    <Window.KeyBindings>
        <KeyBinding Command="{Binding #window.SomeHotKey}" Gesture="Ctrl+S" />
    </Window.KeyBindings>
...
```

```cs
public partial class MainWindow : Window
{
    public MainWindow()
    {
        InitializeComponent();
    }

    public void SomeHotKey()
    {
        Trace.WriteLine("Hit");
    }
}
```

Alternatively, you could use KeyDown events in code-behind for everything. Communication across controls/VMs can be managed via messages if they aren't coupled.

**Reply by cornerbowlsoftware** [reply]
OK nice syntax. I didn't realize I could use #window.

**Reply by stevemonaco** [reply]
It requires a named control in the same scope, so basically within the same file (not including `DataTemplate`s). See https://docs.avaloniaui.net/docs/guides/data-binding/binding-to-controls#binding-to-a-named-control

**Comment by cornerbowlsoftware** [comment]
@stevemonaco I gave this a shot as follows:

```
<local:BaseDataGrid
    x:Class="ServerManagerAvaloniaApp.FlatDataGrid"
    xmlns="https://github.com/avaloniaui"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:local="using:ServerManagerAvaloniaApp"
    xmlns:lib="using:ServerManagerLibrary"
    mc:Ignorable="d"
    d:DesignHeight="200" d:DesignWidth="400"
    x:DataType="lib:BaseViewModel"
    x:Name="baseDataGrid">

	<DataGrid.KeyBindings>
		<KeyBinding Gesture="Ctrl+C" Command="{Binding Copy}" />
		<KeyBinding Gesture="Delete" Command="{Binding Delete}" />
		<KeyBinding Gesture="Ctrl+V" Command="{Binding #baseDataGrid.Paste}" />
	</DataGrid.KeyBindings>
	
	<DataGrid.ContextMenu>
		<ContextMenu ItemsSource="{Binding MenuItemVMs}">		
		</ContextMenu>		
	</DataGrid.ContextMenu>

</local:BaseDataGrid>
```
When pressing Ctrl+V my handler is not triggered and instead the app crashes with the following exception:

`System.InvalidProgramException: 'Common Language Runtime detected an invalid program.'`

**Reply by cornerbowlsoftware** [reply]
In a new project with just your code inserted the app does not crash but also does not call the function:

```
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             xmlns:vm="clr-namespace:TestKeyBindings.ViewModels"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="TestKeyBindings.Views.MainView"
             x:DataType="vm:MainViewModel"
             x:Name="window">

	<UserControl.KeyBindings>
		<KeyBinding Command="{Binding #window.SomeHotKey}" Gesture="Ctrl+S" />
	</UserControl.KeyBindings>

  <TextBlock Text="{Binding Greeting}" HorizontalAlignment="Center" VerticalAlignment="Center"/>
</UserControl>
```
```
using Avalonia.Controls;

namespace TestKeyBindings.Views;

public partial class MainView : UserControl
{
    public MainView()
    {
        InitializeComponent();
    }

    public void SomeHotKey()
    {
    }
}
```

**Reply by cornerbowlsoftware** [reply]
This only works when applied to a Window which doesn't do me any good since I need the command to be routed to my controls.

**Reply by stevemonaco** [reply]
This is because you have nothing that is focusable. `UserControl` and `TextBlock` aren't focusable by default. Try setting `Focusable="True"` on the `UserControl` and it should work. Though it's not necessary if any controls inside of it were focusable.

There's possibly a more elegant solution here, but I don't know of one.

**Reply by cornerbowlsoftware** [reply]
Great, when I explicitly set the focus to the view I get the call. Thanks.
